### PR TITLE
Dice loss api

### PR DESCRIPTION
**Pull request by fantes**
add api for dice losses 
- "loss":"dice"  
- "loss":"dice_binary" 
- "loss":"dice_binary_weighted"

 in `mllib` (caffelib) parameters

WARNING: only dice is implemented in caffe atm

### CODE REVIEW

**Comment by fantes** [comment]
train_data
['/data1/segmentation/NucleiSegmentation/train_480x480.txt', '/data1/segmentation/NucleiSegmentation/test_480x480.txt']
parameters_input
{'segmentation': True, 'shuffle': True}
parameters_mllib
{'solver': {'test_initialization': False, 'base_lr': 0.001, 'snapshot': 500, 'iterations': 80000, 'test_interval': 500, 'iter_size': 1, 'solver_type': 'SGD'}, 'resume': False, 'gpuid': 0, 'timesteps': 32, 'gpu': True, 'net': {'test_batch_size': 2, 'batch_size': 4}}
parameters_output
{'measure': ['acc', 'meaniou']}

File: src/backends/caffe/caffelib.cc [hunk 1]
```diff
@@ -187,6 +188,153 @@ namespace dd
 	caffe::ReadProtoFromTextFile(dest_deploy_net,&deploy_net_param);
 
 
+    if (this->_loss == 1)
```
Comment (beniz) [inline src/backends/caffe/caffelib.cc]: Can we hold the loss into a string instead, thus paving the way to using more custom losses via API ?

File: src/backends/caffe/caffelib.cc [hunk 2]
```diff
@@ -356,6 +504,68 @@ namespace dd
       throw MLLibBadParamException("error reading or listing Caffe models in repository " + this->_mlmodel._repo);
   }
 
+
+  template <class TInputConnectorStrategy, class TOutputConnectorStrategy, class TMLModel>
+  caffe::LayerParameter*
+  CaffeLib<TInputConnectorStrategy,TOutputConnectorStrategy,TMLModel>::insert_layer_before(caffe::NetParameter &net_param, int layer_number)
```
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/caffe/caffelib.cc]: agreed :)

File: src/backends/caffe/caffelib.cc [hunk 3]
```diff
@@ -187,6 +188,153 @@ namespace dd
 	caffe::ReadProtoFromTextFile(dest_deploy_net,&deploy_net_param);
 
 
+    if (this->_loss == 1)
+      // dice loss!!
+      {
```
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/caffe/caffelib.cc]: done

File: src/backends/caffe/caffelib.cc (hunk 1 above)
Comment (fantes) [reply src/backends/caffe/caffelib.cc]: done

File: tests/ut-caffe-mlp.cc [hunk 4]
```diff
@@ -605,3 +605,69 @@ TEST(caffelib,configure_resnet_template_n_nt)
   ASSERT_EQ(276,deploy_net_param.layer_size());
   delete caff;
 }
+
+TEST(caffelib, configure_deeplabvgg16_diceloss)
```
[the remark this answers is not in the document]
Comment (fantes) [reply tests/ut-caffe-mlp.cc]: its does not test much, but allows to output modified prototxt for visual inspection

File: src/backends/caffe/caffelib.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/caffe/caffelib.cc]: there are also some function "find" by name, by type, etc...

**Comment by beniz** [comment]
Can you rebase to master please ? (https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request). Master now has new continuous integration scheme. Thanks.

**Comment by beniz** [comment]
PR test fails on `./ut_caffe_mlp` with error:

```
unknown file: Failure
C++ exception with description "logger with name UT already exists" thrown in the test body.
```

This is because spdlog uses static logger names, and thus you need to rename a logger for every test, e.g.
```
caff->_logger = spdlog::stdout_logger_mt("test2");
```

**Comment by fantes** [comment]
```
curl -X PUT "http://localhost:8080/services/imageserv" -d '{
       "mllib":"caffe",
       "description":"image detection service",
       "type":"supervised",
       "parameters":{
         "input":{                              
           "connector":"image","bbox":true
         },                                     
         "mllib":{
           "nclasses":21,"weights":"VGG.caffemodel","finetuning":true,"template":"ssd_300"                                        
         }                      
       },       
       "model":{                           
         "templates":"../templates/caffe/",
         "repository":"/path/to/data/"
       }}'
```
curl -X POST "http://localhost:8080/train" -d '{
       "service":"imageserv",
       "async":true,
       "parameters":{
         "mllib":{
           "gpu":true,"bbox":true,"resume":false,
           "net":{
             "batch_size":24,"test_batch_size":1
           },
           "solver":{
             "test_interval":1000,"snapshot":1000,
             "iterations":70000,
             "base_lr":0.0001,
             "solver_type":"RMSPROP"
           },"distort":{"all_effects":true,"prob":0.5}
         },
         "input":{
           "connector":
           "image",
           "width":300,
           "height":300
         },
         "output":{
           "measure":["map"]
         }
       },
       "data":["/path/to/train.txt","/path/to/test.txt"]
}'

**Pull request merged**